classes/webservice/ApiNode.php

```diff
@@ -0,0 +1,358 @@
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+class ApiNode
+{
+    public const TYPE_VALUE = 'value';
+    public const TYPE_LANGUAGE = 'language';
+    public const TYPE_PARENT = 'parent';
+    public const TYPE_LIST = 'list';
+
+    public static $languages;
+    private $type;
+    private $name;
+    private $value;
+    private $attributes;
+    private $nodes;
+
+    private function __construct($type, $name = null, $value = null, $attributes = [], $nodes = [])
+    {
+        $this->type = $type;
+        $this->name = $name;
+        $this->value = $value;
+        $this->attributes = $attributes;
+        $this->nodes = $nodes;
+    }
+
+    /**
+     * Create new ApiNode instance of type "value"
+     *
+     * @param string $name
+     * @param string $value
+     *
+     * @return \ApiNode
+     */
+    private static function value($name, $value = null)
+    {
+        return new ApiNode(self::TYPE_VALUE, $name, $value);
+    }
+
+    /**
+     * Create new ApiNode instance of type "lang"
+     * Lang ApiNode serves as ApiNode with name. All the translated values are supposed to be children of that Node.
+     *
+     * @param string $name
+     *
+     * @return \ApiNode
+     */
+    private static function lang($name)
+    {
+        return new ApiNode(self::TYPE_LANGUAGE, $name);
+    }
+
+    /**
+     * Create new ApiNode instance of type "parent"
+     * Parent ApiNode serves as ApiNode with name and array of child nodes (and potentionally array of attributes)
+     * Its children nodes are meant to be rendered as associative arrays
+     *
+     * @param string $name
+     * @param array $attributes
+     *
+     * @return \ApiNode
+     */
+    public static function parent($name = null, $attributes = [])
+    {
+        return new ApiNode(self::TYPE_PARENT, $name, null, $attributes);
+    }
+
+    /**
+     * Create new ApiNode instance of type "list"
+     * List ApiNode serves as ApiNode with name and array of child nodes.
+     * Its children nodes are meant to be rendered as non-associative arrays
+     *
+     * @param string $name
+     * @param array $attributes
+     *
+     * @return \ApiNode
+     */
+    public static function list($name = null, $attributes = [])
+    {
+        return new ApiNode(self::TYPE_LIST, $name, null, $attributes);
+    }
+
+    /** @return string */
+    public function getType()
+    {
+        return $this->type;
+    }
+
+    /** @return string */
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    /** @return string */
+    public function getValue()
+    {
+        return $this->value;
+    }
+
+    /** @return array|null */
+    public function getAttributes()
```

Comment:
```suggestion
    public function getAttributes(): ?attributes
```

Reply:
Done in b4d8238 - with one exception - default value for `$attributes` and `$nodes` unified as empty array instead of null. (Adding operations gets easier as there is no need to check whether `$attributes` or `$nodes` ar array or null)